### DISCUSSION TITLE
[28.10.2024] Blog 3 - Verhaltensmodellierung

### DISCUSSION THREAD

**Discussion by TimHitschfeld** [bunki1224/Softwareengineering-Projekt#27] [Show and tell]
Die Planung und Umsetzung einer Reiseplaner-Webapp erfordert eine klare Modellierung der verschiedenen Interaktionen und Prozesse, die während der Nutzung der App auftreten. In diesem Beitrag werfen wir einen Blick auf die Verhaltensmodellierung dieses Projekts und erklären die Funktionen und Abläufe anhand von vier Hauptdiagrammen: Sequenzdiagramme und Aktivitätsdiagramme, die die Benutzerführung, das Abrufen von Orten sowie das Anlegen und Planen von Aktivitäten visualisieren.

1. Benutzerregistrierung und Validierung
![User-Registration](https://github.com/user-attachments/assets/07e8bfb3-feaa-430c-b76d-c3519547d7f7)

Im ersten Schritt der Nutzung unserer Reiseplaner-App muss sich der Benutzer registrieren. Das Sequenzdiagramm zur Benutzerregistrierung illustriert den Ablauf dieses Prozesses und stellt sicher, dass nur einzigartige Benutzernamen verwendet werden:

Frontend: Der Benutzer füllt das Registrierungsformular aus und initiiert die Prüfung der Verfügbarkeit des Benutzernamens.
Server/API: Die Anfrage zur Verfügbarkeitsprüfung wird an die API weitergeleitet, die dann die Datenbank überprüft.
Datenbank: Es wird geprüft, ob der Benutzername bereits existiert. Ist der Benutzername verfügbar, wird dies an das Frontend zurückgemeldet, andernfalls erfolgt eine Fehlermeldung.
Benutzererstellung: Bei erfolgreicher Überprüfung werden die Daten an den Server geschickt, der den neuen Benutzer in der Datenbank speichert und eine Bestätigung an das Frontend zurückgibt.
Dieser Ablauf stellt sicher, dass Benutzer nur eindeutige Namen verwenden können und direktes Feedback über die Gültigkeit des Benutzernamens erhalten.

2. Ortssuche und Integration von Google Maps API
![Google-Search-Places-Sequenz drawio](https://github.com/user-attachments/assets/dfe12d6a-70e0-45b0-8d55-0f99cc26d546)

Die Webapp bietet eine Funktion zur Ortssuche, um dem Benutzer passende Standorte für geplante Aktivitäten vorzuschlagen. Dieses Sequenzdiagramm illustriert den Ablauf der Kommunikation zwischen der App, dem Google Maps API und der Datenbank:

Frontend: Der Benutzer gibt eine Suchanfrage für einen Ort ein.
Server: Der Server leitet die Anfrage an die Google Maps API weiter, um eine Liste von relevanten Orten zu erhalten.
Google Maps API: Die API sendet die Ergebnisse an den Server zurück.
Speichern des Orts: Der Server speichert den ausgewählten Ort in der Datenbank und bestätigt die Speicherung an das Frontend.
Diese Funktion bietet eine nahtlose Integration von Drittanbieter-Diensten und erleichtert dem Benutzer die Auswahl von Orten.

3.Login und Dashboard
![Aktivitaets-UML](https://github.com/user-attachments/assets/06a299d4-9c70-4fe1-94f3-a7f0be347b27)

Der Benutzer startet die App und wird direkt zur Login-Seite weitergeleitet, sofern er noch nicht eingeloggt ist. Nach einer erfolgreichen Anmeldung gelangt er zum Trip-Dashboard, wo alle erstellten Trips übersichtlich angezeigt werden. Falls der Login fehlschlägt, wird der Benutzer erneut zur Login-Seite weitergeleitet, um seine Anmeldedaten zu überprüfen. Das Dashboard dient als zentrale Anlaufstelle für die Verwaltung und Auswahl bestehender Trips sowie für die Erstellung neuer Reisepläne.

4.Trip-Erstellung und Aktivitätenplanung
Innerhalb des Dashboards kann der Benutzer entweder einen bereits bestehenden Trip auswählen und bearbeiten oder einen komplett neuen Trip anlegen. In diesem Bereich kann der Benutzer Aktivitäten hinzufügen und diese detailliert planen, indem er relevante Informationen wie Uhrzeiten, Orte und Details zu jeder Aktivität einträgt. Die Webapp ermöglicht es, die geplanten Aktivitäten auf einem Zeitstrahl anzuordnen, um eine klare Übersicht über den Tagesablauf zu schaffen. Diese Funktion stellt sicher, dass der Benutzer den gesamten Reiseverlauf optimal strukturieren und Anpassungen vornehmen kann.

5.Export und Nutzung des Reiseplans
Sobald alle Aktivitäten geplant und im Zeitstrahl arrangiert sind, kann der Benutzer den kompletten Reiseplan exportieren.  Der Benutzer hat somit die Möglichkeit, den Plan auch offline zu verwenden, was besonders nützlich für Reisen ohne konstante Internetverbindung ist. Die exportierte Version bietet alle relevanten Details und erleichtert die Umsetzung der geplanten Reise.

Die Diagramme wurden bei den Softwareanforderungsspezifikation [ergänzt.](https://docs.google.com/document/d/1NcsYgDIPpqtPlBKg4xABdPuX1W0yIMI3/edit?usp=sharing&ouid=115995417551287124641&rtpof=true&sd=true)

Euer TripAhead Team (TH)

**Comment by themsgcentre** [comment]
Besonders gefällt uns, dass ihr in den Aktivitätsdiagrammen Gruppen von zusammenhängenden Aktionen in Aktivitätsblöcke untergliedert. Dies hilft dabei, den Prozess und die einzelnen Teilschritte nachzuvollziehen! Auch eure Sequenzdiagramme illustrieren das Zusammenspiel eurer verschiedenen Systeme sehr gut und fördern das Verständnis für die geplante Architektur. Die zusätzliche Beschreibung der Diagramme finden wir ebenfalls sehr gelungen.

Allerdings fänden wir es noch hilfreich, den Nutzer bei den Sequenzdiagrammen ebenfalls als Objekt hinzuzufügen. Dadurch würde die Interaktion aus Nutzersicht noch verständlicher werden!

Insgesamt eine bemerkenswerte Entwicklung, und wir sind gespannt auf die weiteren Updates.

Viele Grüße,
Projekt TDC (HM)

**Comment by MB-stack06** [comment]
Ein sehr informativer Beitrag, das GymTrackerX-Team freut sich jede Woche erneut, euren fortschritt und Ideenreichtum nach zu verfolgen! Eure art und weiße, von der Verwendung von Sequenz- und Aktivitätsdiagrammen zur Modellierung der Abläufe in der Reiseplaner-Webapp gibt einen klaren Einblick in die einzelnen Prozessschritte und die Interaktion der verschiedenen Komponenten. Besonders die anschauliche Darstellung der Benutzerregistrierung und Ortssuche verdeutlicht, wie die App funktioniert und wie nahtlos Frontend, Server und Datenbank miteinander kommunizieren. Die Diagramme machen die Abläufe gut nachvollziehbar und unterstreichen, dass eine benutzerfreundliche und strukturierte Reiseplanung hier im Mittelpunkt steht. Danke an das TripAhead-Team für diesen aufschlussreichen Einblick!
Wir behalten euren Fortschritt weiter im Auge und denken, dass jeden noch viel von eurer Arbeit lernen kann.

Euer GymTrackerX-Team.

**Comment by schmid-di** [comment]
Wir schätzen, dass ihr in eurem Aktivitätsdiagramm die verschiedenen Funktionsbereiche wie Login, Trip-Erstellung und Aktivitätenplanung klar voneinander abgrenzt habt. Die Unterteilung in diese Blöcke macht den Ablauf nachvollziehbar und gibt Struktur. Allerdings könnten die einzelnen Schritte innerhalb dieser Blöcke noch etwas detaillierter dargestellt werden, um besser zu verstehen, wie die Aktionen ineinandergreifen.

Auch die Beschreibung der Diagramme im Blog ist hilfreich, um die Architektur der App zu verstehen. Dennoch würden wir uns wünschen, dass einige Bereiche, wie die Benutzerführung im Trip-Dashboard, noch genauer erläutert werden. Gerade bei der Darstellung des Zusammenspiels der einzelnen Systeme wäre eine präzisere Darstellung sinnvoll, damit die Struktur der App noch klarer wird.

Ein weiterer Punkt: Es wäre aus unserer Sicht hilfreich, den Benutzer als eigenes Objekt im Aktivitätsdiagramm aufzuführen. So würde die Interaktion aus Nutzersicht transparenter, was besonders für die User Experience wichtig ist.

Insgesamt seid ihr auf einem guten Weg, und die Modellierung zeigt viel Potenzial. Ein paar zusätzliche Details würden die Klarheit jedoch noch steigern.

Viele Grüße
Euer UMOC Team

/JM /PL
